hal/explicit_pinmap.h

```diff
@@ -0,0 +1,183 @@
+ * limitations under the License.
+ */
+#ifndef EXPLICIT_PINMAP_H
+#define EXPLICIT_PINMAP_H
+#define CONSTEXPR constexpr
+#include "PeripheralPinMaps.h"
+
+constexpr PinMap get_pwm_pinmap_t(const PinName pin)
+{
+    for (uint32_t i = 0; i < (sizeof(PINMAP_PWM)/sizeof(PINMAP_PWM[0])); i++) {
+        if (PINMAP_PWM[i].pin == pin) {
+            return {pin, PINMAP_PWM[i].peripheral, PINMAP_PWM[i].function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+constexpr PinMap get_analogin_pinmap_t(const PinName pin)
+{
+    for (uint32_t i = 0; i < (sizeof(PINMAP_ANALOGIN)/sizeof(PINMAP_ANALOGIN[0])); i++) {
+        if (PINMAP_ANALOGIN[i].pin == pin) {
+            return {pin, PINMAP_ANALOGIN[i].peripheral, PINMAP_ANALOGIN[i].function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+constexpr PinMap get_analogout_pinmap_t(const PinName pin)
+{
+    for (uint32_t i = 0; i < (sizeof(PINMAP_ANALOGOUT)/sizeof(PINMAP_ANALOGOUT[0])); i++) {
+        if (PINMAP_ANALOGOUT[i].pin == pin) {
+            return {pin, PINMAP_ANALOGOUT[i].peripheral, PINMAP_ANALOGOUT[i].function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+constexpr i2c_pinmap_t get_i2c_pinmap_t(const PinName sda, const PinName scl)
+{
+    int sda_idx = NC;
+    int scl_idx = NC;
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_I2C_SDA)/sizeof(PINMAP_I2C_SDA[0])); i++) {
+        if (PINMAP_I2C_SDA[i].pin == sda) {
+            sda_idx = i;
+            break;
+        }
+    }
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_I2C_SCL)/sizeof(PINMAP_I2C_SCL[0])); i++) {
+        if (PINMAP_I2C_SCL[i].pin == scl) {
+            scl_idx = i;
+            break;
+        }
+    }
+
+    if (sda_idx == NC || scl_idx == NC || PINMAP_I2C_SDA[sda_idx].peripheral != PINMAP_I2C_SCL[scl_idx].peripheral) {
+        return {NC, NC, NC, NC, NC};
+    }
+
+    return {PINMAP_I2C_SDA[sda_idx].peripheral, PINMAP_I2C_SDA[sda_idx].pin, PINMAP_I2C_SDA[sda_idx].function, PINMAP_I2C_SCL[scl_idx].pin, PINMAP_I2C_SCL[scl_idx].function};
+}
+
+constexpr serial_pinmap_t get_uart_pinmap_t(const PinName tx, const PinName rx)
+{
+    int tx_idx = NC;
+    int rx_idx = NC;
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_UART_TX)/sizeof(PINMAP_UART_TX[0])); i++) {
+        if (PINMAP_UART_TX[i].pin == tx) {
+            tx_idx = i;
+            break;
+        }
+    }
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_UART_RX)/sizeof(PINMAP_UART_RX[0])); i++) {
+        if (PINMAP_UART_RX[i].pin == rx) {
+            rx_idx = i;
+            break;
+        }
+    }
+
+    if (tx_idx == NC || rx_idx == NC || PINMAP_UART_TX[tx_idx].peripheral != PINMAP_UART_RX[rx_idx].peripheral) {
+        return {NC, NC, NC, NC, NC};
+    }
+
+    return {PINMAP_UART_TX[tx_idx].peripheral, PINMAP_UART_TX[tx_idx].pin, PINMAP_UART_TX[tx_idx].function, PINMAP_UART_RX[rx_idx].pin, PINMAP_UART_RX[rx_idx].function};
+}
+
+constexpr serial_fc_pinmap_t get_uart_fc_pinmap_t(const PinName rxflow, const PinName txflow)
```

Comment:
Still think these functions shouldn't have the `_t` suffix. You're getting the `pinmap`, which has type `pinmap_t`. Only types should have the `_t` suffix.

Reply:
Fixed.